### PR TITLE
Read current term register asynchronously in StoreHeartbeatService

### PR DESCRIPTION
**Pull request by fcofdez**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/stateless/StoreHeartbeatService.java [hunk 1]
```diff
@@ -19,18 +19,19 @@
...
 import org.elasticsearch.threadpool.ThreadPool;
 
-import java.util.function.LongSupplier;
+import java.util.OptionalLong;
+import java.util.function.Consumer;
 
 public class StoreHeartbeatService implements LeaderHeartbeatService {
     public static final Setting<TimeValue> HEARTBEAT_FREQUENCY = Setting.timeSetting(
-        "heartbeat_frequency",
+        "coordination.stateless.heartbeat_frequency",
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/coordination/stateless/StoreHeartbeatService.java:27]: That's why I shouldn't code late... 🤦

File: server/src/main/java/org/elasticsearch/cluster/coordination/stateless/StoreHeartbeatService.java [hunk 2]
```diff
@@ -134,13 +138,22 @@ protected void doRun() throws Exception {
...
-            final var registerTerm = currentTermSupplier.getAsLong();
-            if (registerTerm == heartbeatTerm) {
-                heartbeatStore.writeHeartbeat(new Heartbeat(heartbeatTerm, absoluteTimeInMillis()), rerunListener);
-            } else {
-                assert heartbeatTerm < registerTerm;
-                listener.onResponse(registerTerm);
-            }
+            currentTermSupplier.accept(rerunListener.delegateFailure((delegate, registerTermOpt) -> {
+                if (registerTermOpt.isEmpty()) {
+                    rerunListener.onResponse(retryAfterTermReadFailureDelay);
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/cluster/coordination/stateless/StoreHeartbeatService.java:143]: I think it's fine to not cap the max number of retries here, if we fail to write the heartbeat a few times some other node should take over.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by fcofdez** [comment]
Thanks David!

**Pull request merged**